### PR TITLE
Get and set material constants in GUI scripts via gui.get/gui.set

### LINKED ISSUE #9005: Add support for setting/getting constants via gui.get/gui.set/gui.animate

### PR DESCRIPTION
**Pull request by Jhonnyg**
`gui.get` and `gui.set` can now be used to get or set material constants for a GUI node:
```lua
local node = gui.get_node("box")
local tint = gui.get(node, "tint")
tint.x = tint.x + dt * 0.1
go.set(node, "tint", tint)
```

Fixes #9005

### CODE REVIEW

File: engine/gamesys/src/waf_gamesys.py [hunk 1]
```diff
@@ -303,10 +303,15 @@ def transform_gui(task, msg):
...
     texture_names = set()
+    material_names = set()
+
     msg.material = msg.material.replace(".material", ".materialc")
     for f in msg.fonts:
         font_names.add(f.name)
         f.font = f.font.replace('.font', '.fontc')
+    for f in msg.materials:
+        material_names.add(f.name)
+        f.material = f.material.replace('.material', '.materialc')
```
Comment (Jhonnyg) [inline engine/gamesys/src/waf_gamesys.py:314]: Could we remove waf_gamesys.py and use bob for building test content? Feels really strange that we have yet another way of building content.. :(

File: engine/gameobject/src/gameobject/gameobject_script.cpp [hunk 2]
```diff
@@ -524,72 +524,6 @@ namespace dmGameObject
         return RESULT_OK;
     }
 
-    static int CheckGoGetResult(lua_State* L, dmGameObject::PropertyResult result, const PropertyDesc& property_desc, dmhash_t property_id, dmGameObject::HInstance target_instance, const dmMessage::URL& target, const dmGameObject::PropertyOptions& property_options, bool index_requested)
```
Comment (Jhonnyg) [inline engine/gameobject/src/gameobject/gameobject_script.cpp:527]: Moved to gameobject_script_util.cpp so we can share with GUI

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 3]
```diff
@@ -136,6 +138,146 @@ namespace dmGameSystem
         }
     }
 
+    struct RenderGuiContext
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:141]: Some code got moved up so we can reuse it

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 4]
```diff
@@ -1211,24 +1329,29 @@ namespace dmGameSystem
...
             dmParticle::RenderConstant* c = &first_emitter_render_data->m_RenderConstants[i];
-            dmGameSystem::SetRenderConstant(render_constants, c->m_NameHash, (dmVMath::Vector4*) &c->m_Value, c->m_IsMatrix4 ? 4 : 1);
+            dmGameSystem::SetRenderConstant(emitter_render_constants, c->m_NameHash, (dmVMath::Vector4*) &c->m_Value, c->m_IsMatrix4 ? 4 : 1);
         }
 
-        if (render_constants) {
-            dmGameSystem::EnableRenderObjectConstants(&ro, render_constants);
+        // NOTE: This means that if any of the particle FX emitters has render constants, they will take precedence over the render cosntants
+        //       that has been set on the node via go.set.
+        HComponentRenderConstants render_constants_to_use = emitter_render_constants ? emitter_render_constants : render_constants;
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:1350]: I _think_ this is correct? If someone used particlefx.set_constant on a specific emitter, it should probably take precedence over someone who has set the constant on a node via gui.set 🤔

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 5]
```diff
@@ -787,6 +929,11 @@ namespace dmGameSystem
...
         scene_params.m_GetResourceCallback = GetSceneResourceByHash;
         scene_params.m_GetResourceCallbackContext = gui_component;
+        scene_params.m_GetMaterialPropertyCallback = GetMaterialPropertyCallback;
+        scene_params.m_GetMaterialPropertyCallbackContext = gui_component;
+        scene_params.m_SetMaterialPropertyCallback = SetMaterialPropertyCallback;
+        scene_params.m_SetMaterialPropertyCallbackContext = gui_component;
+        scene_params.m_DestroyRenderConstantsCallback = DestroyRenderConstantsCallback;
         scene_params.m_OnWindowResizeCallback = &OnWindowResizeCallback;
         scene_params.m_ScriptWorld = gui_world->m_ScriptWorld;
         gui_component->m_Scene = dmGui::NewScene(scene_resource->m_GuiContext, &scene_params);
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:939]: The context for all of these callbacks are the same, do we _really_ need anything other than the gui_component as the context?

**Review by JCash — DISMISSED** [review]
Lgtm

**Pull request merged**